src/plot_api/plot_api.js

```diff
@@ -2508,6 +2526,314 @@ Plotly.relayout = function relayout(gd, astr, val) {
 };
 
 /**
+ * Transition to a set of new data and layout properties
+ *
+ * @param {string id or DOM element} gd
+ *      the id or DOM element of the graph container div
+ */
+Plotly.transition = function(gd, data, layout, traceIndices, transitionConfig) {
+    gd = getGraphDiv(gd);
+
+    var i, traceIdx;
+    var fullLayout = gd._fullLayout;
+
+    transitionConfig = Lib.extendFlat({
+        ease: 'cubic-in-out',
+        duration: 500,
+        delay: 0,
+        cascade: 0
+    }, transitionConfig || {});
+
+    // Create a single transition to be passed around:
+    if(transitionConfig.duration > 0) {
+        gd._currentTransition = d3.transition()
+            .duration(transitionConfig.duration)
+            .delay(transitionConfig.delay)
+            .ease(transitionConfig.ease);
+    } else {
+        gd._currentTransition = null;
+    }
+
+    var dataLength = Array.isArray(data) ? data.length : 0;
+
+    // Select which traces will be updated:
+    if(isNumeric(traceIndices)) traceIndices = [traceIndices];
+    else if(!Array.isArray(traceIndices) || !traceIndices.length) {
+        traceIndices = gd._fullData.map(function(v, i) {return i;});
```

Comment:
FYI: this is how `restyle` currently handles this -> 

https://github.com/plotly/plotly.js/blob/5b92305206ec84df05fddda65b7f4033170b3bb7/src/plot_api/plot_api.js#L1561-L1564